packages/@aws-cdk/core/lib/private/cloudformation-lang.ts

```diff
@@ -97,44 +47,227 @@ export class CloudFormationLang {
+ *
+ *   Type hinting could have been added to the `IResolvable` protocol as well,
+ *   but for now we just use the type of an encoded value as a type hint. That way
+ *   we don't need to annotate anything more at the L1 level--we will use the type
+ *   encodings added by construct authors at the L2 levels. L1 users can escape the
+ *   default decision of "string" by using `Token.asList()`.
+ *
+ * - COMPLEX KEYS: since tokens can be string-encoded, we can use string-encoded tokens
+ *   as the keys in JavaScript objects. However, after resolution, those string-encoded
+ *   tokens could resolve to intrinsics (`{ Ref: ... }`), which CANNOT be stored in
+ *   JavaScript objects anymore.
+ *
+ *   We therefore need a protocol to store the resolved values somewhere in the JavaScript
+ *   type model,  which can be returned by `resolve()`, and interpreted by `tokenAwareStringify()`
+ *   to produce the correct JSON.
+ *
+ *   And example will quickly show the point:
+ *
+ *    User writes:
+ *       { [resource.resourceName]: 'SomeValue' }
+ *    ------ string actually looks like ------>
+ *       { '${Token[1234]}': 'SomeValue' }
+ *    ------ resolve ------->
+ *       { '$IntrinsicKey$0': [ {Ref: Resource}, 'SomeValue' ] }
+ *    ------ tokenAwareStringify ------->
+ *       '{ "' ++ { Ref: Resource } ++ '": "SomeValue" }'
  */
-function deepQuoteStringsForJSON(x: any): any {
-  if (typeof x === 'string') {
-    // Whenever we escape a string we strip off the outermost quotes
-    // since we're already in a quoted context.
-    const stringified = JSON.stringify(x);
-    return stringified.substring(1, stringified.length - 1);
+function tokenAwareStringify(root: any, space: number, ctx: IResolveContext) {
+  let indent = 0;
+
+  const ret = new Array<Segment>();
+
+  // First completely resolve the tree, then encode to JSON while respecting the type
+  // hints we got for the resolved intrinsics.
+  recurse(ctx.resolve(root, { allowIntrinsicKeys: true }));
+
+  switch (ret.length) {
+    case 0: return undefined;
+    case 1: return renderSegment(ret[0]);
+    default:
+      return fnJoinConcat(ret.map(renderSegment));
   }
 
-  if (Array.isArray(x)) {
-    return x.map(deepQuoteStringsForJSON);
+  /**
+   * Stringify a JSON element
+   */
+  function recurse(obj: any): void {
+    if (obj === undefined) { return; }
+
+    if (Token.isUnresolved(obj)) {
+      throw new Error('This shouldnt happen anymore');
+    }
+    if (Array.isArray(obj)) {
+      return renderCollection('[', ']', obj, recurse);
+    }
+    if (typeof obj === 'object' && obj != null && !(obj instanceof Date)) {
+      // Treat as an intrinsic if this LOOKS like a CFN intrinsic (`{ Ref: ... }`)
+      // AND it's the result of a token resolution. Otherwise, we just treat this
+      // value as a regular old JSON object (that happens to look a lot like an intrinsic).
+      if (isIntrinsic(obj) && resolvedTypeHint(obj)) {
```

Comment:
doesn't `isIntrinsic()` impliy `resolvedTypeHint()` always returns true? Maybe just assert?

Reply:
We can have type hints on things that aren't intrinsics (for example, as returned by a lazy).

We can also have structures that LOOK like intrinsics but shouldn't be treated as such (for example, the literal structure `{ Ref: X }`)

Reply:
> We can have type hints on things that aren't intrinsics (for example, as returned by a lazy).

This conditional won't cover non-intrinsics (it's an `&&`).

> We can also have structures that LOOK like intrinsics but shouldn't be treated as such (for example, the literal structure { Ref: X })
 
Shouldn't `isIntrinsic()` cover this?